proof/refine/ARM/Finalise_R.thy

```diff
@@ -2945,32 +2955,99 @@ lemma schedContextMaybeUnbindNtfn_obj_at'_ntfnSc:
   apply (wpsimp wp: setThreadState_Inactive_unlive set_reply'.set_wp gts_wp' getReplyTCB_wp)
   by (auto simp: ko_wp_at'_def obj_at'_def projectKOs is_aligned_def ps_clear_def
                  objBitsKO_def live'_def live_reply'_def weak_sch_act_wf_def pred_tcb_at'_def)
 
-lemma cleanReply_makes_unlive:
-  "\<lbrace>obj_at' (\<lambda>reply. replyTCB reply = None) rptr\<rbrace>
+lemma cleanReply_obj_at_next_prev_none:
+  "\<lbrace>K (rptr' = rptr)\<rbrace>
    cleanReply rptr
-   \<lbrace>\<lambda>_. ko_wp_at' (Not \<circ> live') rptr\<rbrace>"
+   \<lbrace>\<lambda>_. obj_at' (\<lambda>reply. \<not> bound (replyNext reply) \<and> \<not> bound (replyPrev reply)) rptr'\<rbrace>"
   apply (simp add: cleanReply_def)
   apply (wpsimp wp: set_reply'.set_wp)
-  apply (auto simp: ko_wp_at'_def obj_at'_def projectKOs objBitsKO_def live_reply'_def)
+  apply (auto simp: obj_at'_def projectKOs objBitsKO_def)
   done
 
-lemma replyClear_makes_unlive:
+lemma replyPop_makes_unlive:
   "\<lbrace>\<lambda>s. weak_sch_act_wf (ksSchedulerAction s) s\<rbrace>
+   replyPop rptr tptr
+   \<lbrace>\<lambda>_. ko_wp_at' (Not \<circ> live') rptr\<rbrace>"
+  apply (simp add: replyPop_def)
+  apply (wpsimp wp: replyUnlink_makes_unlive cleanReply_obj_at_next_prev_none
+                    hoare_vcg_if_lift threadGet_wp
+         | wp (once) hoare_drop_imps)+
+  by (clarsimp simp: obj_at'_def)
+
+lemma replyRemove_makes_unlive:
+  "\<lbrace>\<lambda>s. weak_sch_act_wf (ksSchedulerAction s) s\<rbrace>
+   replyRemove rptr tptr
+   \<lbrace>\<lambda>_. ko_wp_at' (Not \<circ> live') rptr\<rbrace>"
+  apply (simp add: replyRemove_def)
+  by (wpsimp wp: replyPop_makes_unlive replyUnlink_makes_unlive cleanReply_obj_at_next_prev_none
+                 hoare_vcg_if_lift threadGet_wp gts_wp' hoare_drop_imps)
+
+lemma replyRemoveTCB_makes_unlive:
+  "\<lbrace>\<lambda>s. st_tcb_at' (\<lambda>st. replyObject st = Some rptr) tptr s
+        \<and> weak_sch_act_wf (ksSchedulerAction s) s\<rbrace>
+   replyRemoveTCB tptr
+   \<lbrace>\<lambda>_. ko_wp_at' (Not \<circ> live') rptr\<rbrace>"
+  apply (simp add: replyRemoveTCB_def)
+  apply (wpsimp wp: replyUnlink_makes_unlive cleanReply_obj_at_next_prev_none
+                    hoare_vcg_if_lift threadGet_wp gts_wp' hoare_drop_imps)
+  by (clarsimp simp: pred_tcb_at'_def obj_at'_def)
+
+method cancelIPC_makes_unlive_hammer =
+  (normalise_obj_at',
+   frule (2) sym_ref_replyTCB_Receive_or_Reply,
+   fastforce simp: weak_sch_act_wf_def pred_tcb_at'_def obj_at'_def projectKOs)
+
+lemma cancelIPC_makes_unlive:
+  "\<lbrace>\<lambda>s. obj_at' (\<lambda>reply. replyTCB reply = Some tptr) rptr s
+        \<and> weak_sch_act_wf (ksSchedulerAction s) s \<and> valid_replies' s\<rbrace>
+   cancelIPC  tptr
+   \<lbrace>\<lambda>_. ko_wp_at' (Not \<circ> live') rptr\<rbrace>"
+  apply (simp add: cancelIPC_def Let_def getBlockingObject_def sym_refs_asrt_def)
+  apply (rule hoare_seq_ext[OF _ stateAssert_sp])
```

Comment:
Maybe worth replacing these two lines with an `apply (intro hoare_seq_ext[OF _ stateAssert_sp]`? You could lump the following sp rule into that `intro` too